docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,420 @@
+    // etc.
+  }
+}
+```
+
+**Advantages**:
+
+1. No additional connection setup. When Conduit starts a connector process, it
+   establishes a connection. The same connection is used for all communication (
+   e.g. configuring a connector, opening, reading/writing records, etc.)
+2. Connector actions (which are planned for a future milestone) might use the
+   same command-and-reply stream.
+
+**Disadvantages**:
+
+1. A single method for all the operations makes both, the server and client
+   implementation, more complex. In Conduit, a single gRPC method needs to check
+   for the command type and then reply with a response. Then the client (i.e.
+   the connector) needs to check the response type. In case multiple commands
+   are sent, we need ordering guarantees.
+
+#### Option 2: Exposing a gRPC service in Conduit
+
+Conduit exposes a service to work with schemas. Connectors access the service
+and call methods on the service.
+
+For this work, a connector (i.e. clients of the schema service) needs Conduit's
+IP address and the gRPC port. It's safe to assume that in most, if not all, real
+world use cases, Conduit and connectors will be running on the same host, so we
+can assume that the host is `localhost`. The gRPC port can be communicated to
+the connector via an environment variable.
+
+This service is intended to be used by connectors only. To facilitate that,
+Conduit can generate tokens that connectors would use to authenticate with
+Conduit.
+
+The service should run on a random port. In VMs with hardened security, that
+might not be always possible, so it should be possible for the schema service to
+run on a pre-defined port.
+
+A skeleton of the gRPC definition of the service would be:
+
+```protobuf
+syntax = "proto3";
+
+service SchemaService {
+  rpc Create(CreateSchemaRequest) returns (CreateSchemaResponse);
+  rpc Fetch(FetchSchemaRequest) returns (FetchSchemaResponse);
+}
+
+message CreateSchemaRequest {
+  string name = 1;
+  // other fields
+}
+
+message CreateSchemaResponse {
+  string id = 1;
+}
+
+message FetchSchemaRequest {
+  string id = 1;
+}
+message FetchSchemaResponse {}
+```
+
+**Advantages**:
+
+1. Easy to understand: the gRPC methods, together with requests and responses,
+   can easily be understood from a proto file.
+2. An HTTP API for the schema registry can easily be exposed (if needed).
+3. This API can be extended to include other methods that a connector might
+   need (e.g. connector storage)
+
+**Disadvantages**:
+
+1. Changes needed to communicate Conduit's gRPC port to the connector.
+2. Streams are faster that gRPC method calls. However, registering or fetching a
+   schema is an infrequent operation, so this is not a concern for us.
+
+#### Chosen option
+
+**Option 2** is the chosen method since it offers more clarity and the support
+for remote Conduit instances.
+
+## Required changes
+
+### Conduit
+
+Conduit needs to expose a gRPC schema service as explained above. The gRPC
```

Comment:
Isn't the connector protocol meant for services a connector exposes? It's also not really connector specific (can be accessed by processors too).

Reply:
Yeah, it's connector specific, WASM processors should get their own interface, and that one probably won't be through gRPC, since they can directly call into Conduit.

Reply:
Right, processors are out of the equation here. Still, isn't the connector protocol meant for services a connector exposes?

Reply:
I don't see it as only the place for the services the connector exposes. It can contain anything related to the "protocol" that Conduit and the connector need to follow to communicate. This service sounds like it's part of that protocol.

Reply:
Gotcha. I was understanding as a "protocol" that needs to be followed when talking to a connector, and not the other way around too (to Conduit). In that case, then your suggestion definitely makes sense. :+1: